File: Android/Paging-CRUD/app/src/main/java/com/hhp227/paging_crud/api/PostService.kt
package com.hhp227.paging_crud.api

import com.hhp227.paging_crud.model.BasicApiResponse
import com.hhp227.paging_crud.model.ListItem
import com.hhp227.paging_crud.util.InjectorUtils
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface PostService {
    @GET("posts")
    suspend fun getPostList(
        @Query("group_id") groupId: Int,
        @Query("offset") offset: Int,
        @Query("load_size") loadSize: Int
    ): BasicApiResponse<List<ListItem.Post>>

    @POST("post")
    @FormUrlEncoded
    suspend fun addPost(
        @Header("Authorization") apiKey: String,
        @Field("text") text: String,
        @Field("group_id") groupId: Int
    ): BasicApiResponse<Int>

    @POST("post/{post_id}")
    @FormUrlEncoded
    suspend fun removePost(
        @Header("Authorization") apiKey: String,
        @Path("post_id") postId: Int,
        @Field("_METHOD") method: String = "DELETE"
    ): BasicApiResponse<Unit>

    companion object {
        fun create(): PostService {
            return InjectorUtils.provideRetrofit()
                .create(PostService::class.java)
        }
    }
}

File: iOS/Paging-CRUD/Paging-CRUD/API/PostService.swift
//
//  PostService.swift
//  Paging-CRUD
//

import Foundation

final class PostService {
    private let session: URLSession

    private let baseUrl = URL(string: "\(URLs.baseUrl)/")!

    init(session: URLSession = .shared) {
        self.session = session
    }

    func getPostList(groupId: Int, offset: Int, loadSize: Int) async throws -> BasicApiResponse<[ListItem.Post]> {
        var components = URLComponents(url: baseUrl.appendingPathComponent("posts"), resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "group_id", value: String(groupId)),
            URLQueryItem(name: "offset", value: String(offset)),
            URLQueryItem(name: "load_size", value: String(loadSize))
        ]
        let (data, _) = try await session.data(from: components.url!)
        return try JSONDecoder().decode(BasicApiResponse<[ListItem.Post]>.self, from: data)
    }

    func addPost(apiKey: String, text: String, groupId: Int) async throws -> BasicApiResponse<Int> {
        let request = formUrlEncodedRequest(
            url: baseUrl.appendingPathComponent("post"),
            apiKey: apiKey,
            fields: ["text": text, "group_id": String(groupId)]
        )
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(BasicApiResponse<Int>.self, from: data)
    }

    func removePost(apiKey: String, postId: Int) async throws -> BasicApiResponse<Bool> {
        let request = formUrlEncodedRequest(
            url: baseUrl.appendingPathComponent("post/\(postId)"),
            apiKey: apiKey,
            fields: ["_METHOD": "DELETE"]
        )
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(BasicApiResponse<Bool>.self, from: data)
    }

    private func formUrlEncodedRequest(url: URL, apiKey: String, fields: [String: String]) -> URLRequest {
        var request = URLRequest(url: url)
        var allowedCharacters = CharacterSet.urlQueryAllowed
        allowedCharacters.remove(charactersIn: "&=+")
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "Authorization")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = fields
            .map { key, value in
                "\(key)=\(value.addingPercentEncoding(withAllowedCharacters: allowedCharacters) ?? value)"
            }
            .joined(separator: "&")
            .data(using: .utf8)
        return request
    }

    static func create() -> PostService {
        return PostService()
    }
}
